File: Java/src/main/java/Solution901_1000/Solution953.java
package Solution901_1000;

/**
 * @author windows
 */
public class Solution953 {
    public boolean isAlienSorted(String[] words, String order) {
        int[] chars = new int[26];
        for (int i = 0; i < order.length(); i++) {
            chars[order.charAt(i) - 'a'] = i;
        }
        for (int i = 1; i < words.length; i++) {
            int length = words[i].length() <= words[i - 1].length() ? words[i].length() : words[i - 1].length();
            int j = 0;
            for (; j < length; j++) {
                if (chars[words[i- 1].charAt(j) - 'a'] < chars[words[i].charAt(j) - 'a']){
                    break;
                }else if (chars[words[i- 1].charAt(j) - 'a'] > chars[words[i].charAt(j) - 'a']){
                    return false;
                }
            }
            if (j == length && words[i].length() < words[i - 1].length()){
                return false;
            }
        }
        return true;
    }


    public static void main(String[] args) {
        Solution953 solution953 = new Solution953();
    }
}

File: C++/src/Solution901_1000/Solution953.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution953.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;

bool compare(string& basicString, string& basicString1, int order[]) {
  int l1 = basicString.size(), l2 = basicString1.size();
  int len = l1 < l2 ? l1 : l2;
  for (int i = 0; i < len; ++i) {
    if (order[basicString[i]- 'a'] < order[basicString1[i] - 'a']){
      return true;
    }
    if (order[basicString[i]- 'a'] > order[basicString1[i] - 'a']){
      return false;
    }
  }
  return l1 < l2;
}

bool isAlienSorted(vector<string>& words, string order) {
  int n = words.size();
  int chars[26];
  for (int i = 0; i < order.size(); ++i) {
    chars[order[i] - 'a'] = i;
  }

  for (int i = 0; i < n - 1; ++i) {
    if (!compare(words[i], words[i + 1], chars)){
      return false;
    }
  }
  return true;
}

